rfd/0229-agent-inventory-ui.md

```diff
@@ -0,0 +1,250 @@
+be quite cumbersome for users and as a result limits the visibility cluster administrators have into their infrastructure. Exposing
+this information in the Web UI would greatly improve the user experience and make it easier for cluster administrators to do
+things like identify instances still running old versions of Teleport and need to be updated.
+
+## Details
+
+### Implementation
+
+The first version of this feature will primarily focus on adding the existing `tctl inventory` and `tctl bots instances` functionality
+to the Web UI and use most of the same underlying backend API. Both agents and bot instances will be fetched via streams and
+combined into a single paginated response ordered alphabetically by name by leveraging the `MergeStreams` utility.
+
+#### Web Endpoint
+
+A new `/webapi/sites/:site/instances` will be used by the Web UI to fetch a paginated list of instances, this endpoint will support
+various query parameters for pagination and filtering, including:
+
+- `search`: Filter by hostname.
+- `query`: Filter by a predicate language expression.
+- `services`: Filter to only return agents running the specified service(s).
+- `upgraders`: Filter to only return agents using the specified external upgrader(s).
+- `updaterGroup`: Filter to only return agents in the specified updater group.
+- `type`: Filters to return only agents, bot instances, or both. Default/undefined is both.
+
+The response will be a JSON object containing the page of agents requested.
+
+##### Example response:
+
+```json
+{
+  "instances": [
+    {
+      "id": "1",
+      "type": "agent",
+      "agent": {
+        "name": "server1",
+        "version": "16.1.3",
+        "services": ["ssh", "db", "desktop"],
+        "upgrader": {
+          "type": "unit",
+          "version": "v2",
+          "group": "group1"
+        }
+      }
+    },
+    {
+      "id": "2",
+      "type": "agent",
+      "agent": {
+        "name": "kube1",
+        "version": "16.5.4",
+        "services": ["kube"],
+        "upgrader": {
+          "type": "kube",
+          "version": "v2",
+          "group": "group1"
+        }
+      }
+    },
+    {
+      "id": "3",
+      "type": "bot",
+      "bot": {
+        "name": "bot1",
+        "version": "16.1.3",
+        "service": "kubernetes",
+        "health": "Unknown"
+      }
+    },
+    {
+      "id": "4",
+      "type": "agent",
+      "agent": {
+        "name": "agent1",
+        "version": "16.1.3",
+        "services": ["kube"],
+        "upgrader": {
+          "type": "kube",
+          "version": "v2",
+          "group": "group2"
+        }
+      }
+    }
+  ],
+  "totalCount": 4,
+  "startKey": ""
+}
+```
+
+#### RPC Service and Proto messages
+
+```protobuf
+service InventoryService {
+  // ListInstances returns a page of Agents and Bot Instances.
+  rpc ListInstances(ListInstancesRequest) returns (ListInstancesResponse);
```

Comment:
nit: given that we already use the convention of `Instance` meaning "agent instance" in a lot of places, could we instead name this something like `ListUnifiedInstances`?  In addition to disambiguating from any future `ListInstances` API for the `Instance` type, that would also have the upside of mirroring the naming convention used by `ListUnifiedResources`, which is kind of the access resource equivalent of this API.